evennia/prototypes/prototypes.py

```diff
@@ -33,7 +33,7 @@
 _PROTOTYPE_TAG_META_CATEGORY = "db_prototype"
 PROT_FUNCS = {}
 
-_RE_DBREF = re.compile(r"(?<!\$obj\()(#[0-9]+)")
+_RE_DBREF = re.compile(r"\$dbref\((\#[0-9]+)\)")
```

Comment:
This is actually incorrect. This custom regex is not needed if we remove the special case of  converting `#dbref`. So this entire regex should be removed and instead the new `dbref()` protfunc be added in `evennia/prototypes/protfuncs.py` where it can be handled by the normal protfunc parser.

Reply:
Well, this is the way I though of it:

`$obj()` takes `#dbref` or `key` or `alias` and does an object search.
The new `$dbref()` ** **enforces** ** that the passed in param is a `#number`. If it isn't, it becomes an invalid invocation of the protofunc.

At any rate, the implementation of `dbref()` would be identical to `obj()`. It would just be another thin wrapper of `search.search_object()`. 

https://github.com/evennia/evennia/blob/175af0715988f90c7898d612bf02f7d0e9bf635f/evennia/prototypes/protfuncs.py#L309-L318

On top of that, if the function was added, passing `key` and/or `alias` to `dbref()` would work just like `obj()` and there will be no enforcement of passing in a `#number`. Which is what this change is about: using the regex to alias `$dbref(#number)` as `$obj(#dbref)`.

As I mentioned https://github.com/evennia/evennia/issues/1704#issuecomment-433700848, `$dbref` would be a subset of `$obj()`. It would just allow to explicitly *typecast* a `#number` string as a `dbref` so object search would occur.

The regex is just to match `$dbref(#number)` **if and only if** the param matches `#number` and the protfunc invoked is `$dbref()`. Once the match is done, it just needs to call `$obj(#number)` which is the one that searches by `#dbref`, `key`, and `alias`es.

The tests I added show when the match occur (when it doesn't) and which protofunc will be invoked when the match happens (ie. invoke `$obj()` when matched).

https://github.com/evennia/evennia/pull/1716/files#diff-cde206fcfe341b1e066d1546d3666ef9R265

```python
    def test_RE_DBREF(self):
        def check_RE_DBREF(value, expected_value,
                exp_is_match, exp_is_search):
            try:
                mm = protlib._RE_DBREF.match(value)
                ms = protlib._RE_DBREF.search(value)
                sub1 = protlib._RE_DBREF.sub("$obj(\\1)", value)
                assert expected_value == sub1
                assert (exp_is_match and mm) or (not exp_is_match and mm is None)
                assert (exp_is_search and ms) or (not exp_is_search and ms is None)
            except Exception as e:
                self.fail("%r" % ((value, mm, ms, expected_value, exp_is_match, exp_is_search, e),))

         # no match
         check_RE_DBREF('1234', '1234', False, False)
        check_RE_DBREF('#1234', '#1234', False, False)
        check_RE_DBREF('(#1234)', '(#1234)', False, False)
        check_RE_DBREF('obj(#1234)', 'obj(#1234)', False, False)
        check_RE_DBREF('dbref(#1234)', 'dbref(#1234)', False, False)
        check_RE_DBREF('$obj(#1234)', '$obj(#1234)', False, False)

        # does match so parser will invoke $obj()
        check_RE_DBREF('$dbref(#1234)', '$obj(#1234)', True, True)
        
        # no match
        check_RE_DBREF('obj($obj(#1234))', 'obj($obj(#1234))', False, False)

        # does match so parser will invoke $obj()
        check_RE_DBREF('obj($dbref(#1234))', 'obj($obj(#1234))', False, True)

        # no match
        check_RE_DBREF('some #1234 value', 'some #1234 value', False, False)
        check_RE_DBREF('some (#1234) value', 'some (#1234) value', False, False)
        check_RE_DBREF('some obj(#1234) value', 'some obj(#1234) value', False, False)
        check_RE_DBREF('some dbref(#1234) value', 'some dbref(#1234) value', False, False)

        # does match so parser will invoke $obj()
        check_RE_DBREF('some $dbref(#1234) value', 'some $obj(#1234) value', False, True)

        # no match
        check_RE_DBREF('some obj($obj(#1234) value)', 'some obj($obj(#1234) value)', False, False)

        # does match so parser will invoke $obj()
        check_RE_DBREF('some obj($dbref(#1234) value)', 'some obj($obj(#1234) value)', False, True)

        # no match
        check_RE_DBREF('some dbref($obj(#1234) value)', 'some dbref($obj(#1234) value)', False, False)

        # does match so parser will invoke $obj()
        check_RE_DBREF('some dbref($dbref(#1234) value)', 'some dbref($obj(#1234) value)', False, True)

```

The other tests demonstrate that plain text `#number` **never** becomes an object search **unless** it's wrapped by `$dbref()` or `$obj()`. When wrapped, the result is the expected output of `$obj()` (i.e. the successful object search and its string `#dbref` representation:

https://github.com/evennia/evennia/blob/175af0715988f90c7898d612bf02f7d0e9bf635f/evennia/prototypes/protfuncs.py#L317

https://github.com/evennia/evennia/pull/1716/files#diff-cde206fcfe341b1e066d1546d3666ef9R347
```python
       # plain text, no search
        self.assertEqual(protlib.protfunc_parser("obj(#1)", session=self.session), 'obj(#1)') 
        self.assertEqual(protlib.protfunc_parser("$obj(#1)", session=self.session), '#1')
        self.assertEqual(protlib.protfunc_parser("dbref(#1)", session=self.session), 'dbref(#1)') # <<< (there is NO $)

       # search
        self.assertEqual(protlib.protfunc_parser("$dbref(#1)", session=self.session), '#1')

        # Bad function invocation
        self.assertEqual(protlib.protfunc_parser("$badfunc(#1)", session=self.session), '<UNKNOWN>')  

       # plain text, no search
        self.assertEqual(protlib.protfunc_parser("stone(#12345)", session=self.session), 'stone(#12345)')   
        self.assertEqual(protlib.protfunc_parser("#1", session=self.session), '#1') 
        self.assertEqual(protlib.protfunc_parser("#12345", session=self.session), '#12345') 
        self.assertEqual(protlib.protfunc_parser("nothing(#1)", session=self.session), 'nothing(#1)')
        self.assertEqual(protlib.protfunc_parser("(#12345)", session=self.session), '(#12345)

         # search by key
        self.assertEqual(protlib.protfunc_parser("$obj(Char)", session=self.session), '#6')

        # plain text, no search
        self.assertEqual(protlib.protfunc_parser("obj(Char)", session=self.session), 'obj(Char)') 

         # search
        self.assertEqual(protlib.protfunc_parser("$objlist(#1)", session=self.session), ['#1'])

        # plain text, no search
        self.assertEqual(protlib.protfunc_parser("objlist(#1)", session=self.session), 'objlist(#1)')
        self.assertEqual(protlib.protfunc_parser("dbref(Char)", session=self.session), 'dbref(Char)')
```

For the last test, I actually disagree:

The test asserts that it's an incorrect invocation of the protofunc `$dbref()`, as the passed in param is **not** a valid `#dbref`. 

```python
        self.assertEqual(protlib.protfunc_parser("$dbref(Char)", session=self.session), '<UNKNOWN>')
```

Notice that `Char` it's a valid key which returns a successful search `#6` when passed onto`$obj()` (one of the previous tests)

```python
# search by key
        self.assertEqual(protlib.protfunc_parser("$obj(Char)", session=self.session), '#6')
```

One thing I realized as I was writing the tests for `$obj()` and `$dbref()` - to validate that `#number` is **never** converted to `#dbref` automatically - is that none of them never assert that the protofunc was actually invoked. 

For instance, these cases:

```python
        self.assertEqual(protlib.protfunc_parser("$obj(#1)", session=self.session), '#1')
        self.assertEqual(protlib.protfunc_parser("$dbref(#1)", session=self.session), '#1')
        self.assertEqual(protlib.protfunc_parser("#1", session=self.session), '#1') 
```

They all return the **same value** although only the first 2 tests actually execute the `$obj()` function which produces `return "#{}".format(obj.id) `. The last one doesn't perform any search (as expected) and the result is a plain string (that looks like a `#dbref` but it isn't).

WADR, adding `def dbref()` to protofunc will do the same thing as long as `dbref()` function has its own regex to validate that the passed in param is a `#number` that represents a `#dbref` that can be searched by `obj()`

Something along these lines in protofunc.py

```python
def dbref(*args, **kwargs):
    if re.match(r'\#'\d+', args[0]):
        return obj(arg[0])
    else:
        return '<UNKNOWN>' # <<< or whatever constant is appropritate
```

Reply:
Hi @Griatch 

With this last commit dab2e3c5776, I hardened many of the existing cases by asserting that no object search happens unless `$obj()`, `$dbref()` and `$objlist()` are explicitly used as `value`.

I also grouped them so it's easier to digest what they do (no search, search, and bad).

Let me know if you really want to throw away `_RE_DBREF` and add `dbref` to `protfuncs.py`. IMHO, it'll be pretty much the same at the end.